### PR TITLE
Add docs on how to write tests when using Active Storage

### PR DESCRIPTION
**Pull request by ghiculescu**
Adds some missing examples that I found handy:

- Introduces `fixture_file_upload`, and adds an example.
- Adds example of how to make fixtures for Active Storage attachments (inspired by [this answer](https://stackoverflow.com/a/55835955/641293), edited with thanks to @gmcgibbon suggestions).
- Adds example of how to safely clean up files when running tests in parallel (which is the default in a new Rails app)

Tidies up a few things:

- Makes the sections on cleaning up files lower level headers (no need for them to be primary sections).
- Updates the suggested service name for testing to match the one used in a new Rails app.
- Updates the path to delete files from. Rather than hard coding it, just get it from Active Storage.

### CODE REVIEW

File: guides/source/active_storage_overview.md [hunk 1]
```diff
@@ -983,6 +1059,19 @@ class ApplicationSystemTestCase < ActionDispatch::SystemTestCase
...
 ```
 
+If you're using [parallel tests][], you should also configure each process to use its own
+folder for Active Storage:
+
+```ruby
+class ApplicationSystemTestCase < ActionDispatch::SystemTestCase
+  # ...
+  parallelize_setup do |i|
+    ActiveStorage::Blob.service.instance_variable_set(:@root, "#{ActiveStorage::Blob.service.root}-#{i}")
```
Comment (ghiculescu) [inline guides/source/active_storage_overview.md:1069]: `root` is an `attr_reader`: https://github.com/rails/rails/blob/main/activestorage/lib/active_storage/service/disk_service.rb#L12

Making it an `attr_accessor` would be helpful for this, but it might not be a smart thing to do in general.

**Review by gmcgibbon — CHANGES_REQUESTED** [review]
I don't think this is an advisable way to add attachments to fixtures. Looking at Active Storage's test suite, we use [`ActiveStorage::FixtureSet.blob`](https://github.com/rails/rails/blob/b038bb27a6162184faef3af6e1c676a893651d9f/activestorage/lib/active_storage/fixture_set.rb#L63).  Placing the store root on your fixture file root also means files can be _uploaded_ to that directory, which is not a good idea.

I believe the parallelization problem is also sidestepped by keeping the store in a temp folder and uploading each file fixture individually via `ActiveStorage::FixtureSet.blob`.

**Comment by ghiculescu** [comment]
Oh man, how did I manage to miss that class 🤦

**Comment by ghiculescu** [comment]
Hmmmm, I think you still need a separate service. Let's say you just use this:

```
test:
  service: Disk
  root: <%= Rails.root.join("tmp/storage") %>
```

And then you set that fixture, and also follow the current advice in the guide and do this in a teardown: `FileUtils.rm_rf(Rails.root.join('tmp', 'storage'))`

After the first test that runs, your fixture file will be deleted. That means if you have another test later that depends on it, that test won't work. So this will fail:

```ruby
  def test_avatar
    avatar = users(:spongebob).avatar
    assert avatar.attached? # passes
    assert_not_nil avatar.download # raises
  end
```

I think the correct way to get around that is:

- Have a tmp dir for fixture files, which is either never deleted, or deleted once all tests complete (eg. using `parallel_teardown` - there must be a minitest native way of doing the same). Suggestion: `tmp/storage_fixtures`.
- Have a different tmp dir for files that are uploaded during tests, which is deleted after each test in `teardown`. Suggestion: `tmp/storage`.
- Call `ActiveStorage::FixtureSet.blob filename: "spongebob.png", service_name: "test_fixtures"` in the fixture yml file

I tested this approach in the app I extracted these docs from, and it works great. And it's much cleaner than my old approach!

@gmcgibbon Can you please sanity check my thinking?

**Comment by ghiculescu** [comment]
Actually, no, I think you still may need this. Not for fixtures, just for regular tests.

```ruby
  parallelize_setup do |i|
    ActiveStorage::Blob.service.instance_variable_set(:@root, "#{ActiveStorage::Blob.service.root}-#{i}")
  end
```

This is technically possible:

- Lots of CPUs on your test machine
- Test process 1 uploads a file
- Test process 2 completes a test, runs teardown, which runs `FileUtils.rm_rf(Rails.root.join('tmp', 'storage'))`
- Test process 1 tries to read the file it uploaded, eg. `user.attachment.download`. File no longer exists 💥

**Comment by gmcgibbon** [comment]
> After the first test that runs, your fixture file will be deleted.

Hm, yeah that would seems correct, but we don't want tests to depend on each other. Theoretically we could snapshot the service root _before_ test and restore it _after_ test. That might call for a test helper, but it would be the least complicated solution I can think of.

And yes, I think multiple async deletes on the filesystem on a single location _could_ lead to problems. I don't think something trivial like `Rails.root.join("tmp/storage/#{Thread.current.object_id}")` would propagate across processes properly either.

**Comment by ghiculescu** [comment]
> Hm, yeah that would seems correct, but we don't want tests to depend on each other. Theoretically we could snapshot the service root _before_ test and restore it _after_ test. That might call for a test helper, but it would be the least complicated solution I can think of.

Having a second Disk service would also solve this, but I'm guessing you don't like that solution. Crazy idea: what if Active Storage maintained a Disk service for fixtures internally?

> I'm curious if we've received reports about this though, if it an issue.

Well anecdotally, it happened to me. That's what inspired me to check for, and then, update these docs!

**Comment by ghiculescu** [comment]
For the record, `paperclip` has very similar advice for how to run parallel tests, and how to delete fixture files after your tests complete: https://github.com/kreeti/kt-paperclip#testing

**Comment by ghiculescu** [comment]
@gmcgibbon just circling back after some time off, did you have any more thoughts on this one?

**Comment by sedubois** [comment]
@ghiculescu thanks for helping to document this feature. For what it's worth, I tried using `ActiveStorage::FixtureSet.blob` as instructed and then clearing the storage folder after each test:
```ruby
  def teardown
    super
    # remove files uploaded from fixture YML files at the end of the test run
    FileUtils.rm_rf(ActiveStorage::Blob.services.fetch(:public).root)
  end
```

However I still get the dreaded `Minitest::UnexpectedError: ActiveStorage::FileNotFoundError: ActiveStorage::FileNotFoundError` when running all my tests. 

NB: I am not using parallel tests (`parallelize` is commented out).

**Comment by ghiculescu** [comment]
Hmmm. Advice would be different if you're not using parallel tests. `parallelize_teardown` runs at the end of *each test run*, not after every individual test. What you're doing there is deleting all your fixtures after each individual test. So in test #1 they exist, as soon as test #2 runs they are gone.

Can you try using `Minitest.after_run` instead, see http://ruslanledesma.com/2016/07/21/before-all-after-all-in-minitest.html

**Comment by sedubois** [comment]
Thanks, that works!

File: guides/source/active_storage_overview.md [hunk 2]
```diff
@@ -985,51 +1008,141 @@ class ApplicationSystemTestCase < ActionDispatch::SystemTestCase
...
+  # ...
 end
 ```
 
+If you're not running parallel tests, use `Minitest.after_run` or the equivalent for your test framework:
+
+```ruby
+# test_helper.rb
+
+Minitest.after_run do
```
Comment (ghiculescu) [inline guides/source/active_storage_overview.md:1138]: @sedubois how does this look?

Comment (sedubois) [reply guides/source/active_storage_overview.md:1138]: LGTM

**Comment by ghiculescu** [comment]
> > I'm curious if we've received reports about this though, if it an issue.
> 
> Well anecdotally, it happened to me. That's what inspired me to check for, and then, update these docs!

Just came across another example of this on a project I'm helping with. The test fails like this:

```
Errno::ENOENT:         Errno::ENOENT: No such file or directory @ rb_sysopen - /home/runner/work/foo/bar/tmp/storage/me/sf/mesfzivxljk9vtgao1bwu0ow8tyo
            test/controllers/users_controller_test.rb:88:in `block (2 levels) in <class:UsersControllerTest>'
            test/controllers/users_controller_test.rb:84:in `block in <class:UsersControllerTest>'
```

This is what the relevant line does:

```ruby
model.active_storage_attachment.attach(fixture_file_upload("alex.png", "image/png"))
```

And the test suite had this:

```
  teardown do
    FileUtils.rm_rf(ActiveStorage::Blob.service.root)
  end
```

So it seems like the file gets moved to the `tmp/storage` dir, then deleted at some point very soon after, when another test in another parallel process finishes.

Prior to writing this PR my advice would have been "just run the tests again". Now I can link to the advice in this PR! Specifically, the solution is to add

```ruby
class ActionDispatch::IntegrationTest
  parallelize_setup do |i|
    ActiveStorage::Blob.service.root = "#{ActiveStorage::Blob.service.root}-#{i}"
  end
end
```

File: guides/source/active_storage_overview.md [hunk 3]
```diff
@@ -985,51 +1008,141 @@ class ApplicationSystemTestCase < ActionDispatch::SystemTestCase
...
+`teardown` callback.
 
 ```ruby
-module RemoveUploadedFiles
-  def after_teardown
-    super
-    remove_uploaded_files
+module ActionDispatch
+  class IntegrationTest
+    teardown do
```
Comment (gmcgibbon) [inline guides/source/active_storage_overview.md:1045]: This should probably continue be `after_teardown` in case a user defines a teardown block that references an uploaded file. Also I think the example above uses `def after_teardown` when it could be using `after_teardown do` like this one.

File: guides/source/active_storage_overview.md [hunk 4]
```diff
@@ -985,51 +1008,141 @@ class ApplicationSystemTestCase < ActionDispatch::SystemTestCase
 end
 ```
 
+If you're using [parallel tests][] and the Disk service, you should configure each process to use its own
```
Comment (gmcgibbon) [inline guides/source/active_storage_overview.md:1011]:
```suggestion
If you're using [parallel tests][] and the `DiskService`, you should configure each process to use its own
```

File: guides/source/active_storage_overview.md [hunk 5]
```diff
@@ -961,10 +961,33 @@ class Uploader {
 
 NOTE: Using [Direct Uploads](#direct-uploads) can sometimes result in a file that uploads, but never attaches to a record. Consider [purging unattached uploads](#purging-unattached-uploads).
 
-Discarding Files Stored During System Tests
+Testing
 -------------------------------------------
 
-System tests clean up test data by rolling back a transaction. Because destroy
+Use [`fixture_file_upload`][] to test uploading a file in an integration or controller test.
+Rails handles it like any other param.
```
Comment (gmcgibbon) [inline guides/source/active_storage_overview.md:968]:
```suggestion
Rails handles files like any other parameter.
```

File: guides/source/active_storage_overview.md [hunk 6]
```diff
@@ -961,10 +961,33 @@ class Uploader {
 
 NOTE: Using [Direct Uploads](#direct-uploads) can sometimes result in a file that uploads, but never attaches to a record. Consider [purging unattached uploads](#purging-unattached-uploads).
```
Comment (gmcgibbon) [inline guides/source/active_storage_overview.md:962]: Does this link need to be updated now that the headers have changed?

File: guides/source/active_storage_overview.md [hunk 7]
```diff
@@ -961,10 +961,33 @@ class Uploader {
...
 
-System tests clean up test data by rolling back a transaction. Because destroy
+Use [`fixture_file_upload`][] to test uploading a file in an integration or controller test.
+Rails handles it like any other param.
+
+```ruby
+class SignupController < ActionDispatch::IntegrationTest
+  test "signup" do
+    post signup_path, params: {
+      name: "Spongebob",
```
Comment (gmcgibbon) [inline guides/source/active_storage_overview.md:974]: The testing guide uses names of Rails team members and contributors when referencing users. Please use these instead for consistency and to avoid any potential copyright issues.

Comment (ghiculescu) [reply guides/source/active_storage_overview.md:974]: Will do. I was copying the pattern here: https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/actionpack/lib/action_dispatch/testing/test_process.rb#L11 - should I change that too? (I realise that's not in a guide, but it is in public docs)

File: activestorage/lib/active_storage/service/disk_service.rb [hunk 8]
```diff
@@ -9,7 +9,7 @@ module ActiveStorage
   # Wraps a local disk path as an Active Storage service. See ActiveStorage::Service for the generic API
   # documentation that applies to all services.
   class Service::DiskService < Service
-    attr_reader :root
+    attr_accessor :root
```
Comment (gmcgibbon) [inline activestorage/lib/active_storage/service/disk_service.rb:12]: I think this should also be a separate commit to the doc updates.

File: guides/source/active_storage_overview.md (hunk 7 above)
Comment (gmcgibbon) [reply guides/source/active_storage_overview.md:974]: Ah! Interesting. I'm a little more worried about the guides because of SEO. I don't see any policy on this sort of thing in the contributor's guide, but I would accept a change for this line as well.

File: activestorage/lib/active_storage/service/disk_service.rb (hunk 8 above)
Comment (ghiculescu) [reply activestorage/lib/active_storage/service/disk_service.rb:12]: https://github.com/rails/rails/pull/42163

File: guides/source/active_storage_overview.md (hunk 6 above)
Comment (ghiculescu) [reply guides/source/active_storage_overview.md:962]: No, but good pickup. That refers to a separate section that was added in https://github.com/rails/rails/commit/f4c51e8660bf1771268d45325fff8378eb90698e

image

File: guides/source/active_storage_overview.md (hunk 3 above)
Comment (ghiculescu) [reply guides/source/active_storage_overview.md:1045]: Will update. Re the block - `after_teardown` doesn't take one: https://github.com/rails/rails/blob/main/activesupport/lib/active_support/testing/setup_and_teardown.rb#L44

So the correct pattern seems to be to override the method and call `super`.

File: guides/source/active_storage_overview.md (hunk 7 above)
Comment (ghiculescu) [reply guides/source/active_storage_overview.md]: Updated both to be safe and consistent. https://github.com/rails/rails/pull/41543/files#diff-2f684945f008de83109c95e5234ed9c2c1cfc7abbfd55ec148b9a25d4c18c85dR11

File: guides/source/active_storage_overview.md [hunk 9]
```diff
@@ -985,51 +1008,141 @@ class ApplicationSystemTestCase < ActionDispatch::SystemTestCase
...
+  # ...
 end
 ```
 
+If you're not running parallel tests, use `Minitest.after_run` or the equivalent for your test framework:
+
+```ruby
+# test_helper.rb
+
+Minitest.after_run do
```
[the remark this answers is not in the document]
Comment (ghiculescu) [reply guides/source/active_storage_overview.md:1133]: I left it as is, but I also mention RSpec here: https://github.com/rails/rails/pull/41543/files#diff-e61fcdded4b6112b041ff388609c7032d09b8212ce209d67a7debc32f230d77aR1128

File: activestorage/lib/active_storage/service/disk_service.rb (hunk 8 above)
Comment (gmcgibbon) [reply activestorage/lib/active_storage/service/disk_service.rb:12]: I think this can be reverted if we're handling it in another PR.

**Review by gmcgibbon — APPROVED** [review]

File: activestorage/lib/active_storage/service/disk_service.rb (hunk 8 above)
Comment (ghiculescu) [reply activestorage/lib/active_storage/service/disk_service.rb:12]: Yep! I was gonna do that after https://github.com/rails/rails/pull/42163 was merged, since technically the code samples in this PR won't work without this code change.

Comment (ghiculescu) [reply activestorage/lib/active_storage/service/disk_service.rb:12]: But that's overly pedantic, I'll just do it now.

**Comment by ghiculescu** [comment]
I think that test fail is unrelated or intermittent.

File: guides/source/active_storage_overview.md [hunk 10]
```diff
@@ -960,75 +960,184 @@ class Uploader {
...
+
+System tests clean up test data by rolling back a transaction. Because `destroy`
 is never called on an object, the attached files are never cleaned up. If you
 want to clear the files, you can do it in an `after_teardown` callback. Doing it
 here ensures that all connections created during the test are complete and
 you won't receive an error from Active Storage saying it can't find a file.
 
 ```ruby
-class ApplicationSystemTestCase < ActionDispatch::SystemTestCase
+class ActionDispatch::SystemTestCase
```
Comment (gmcgibbon) [inline guides/source/active_storage_overview.md:996]: Sorry, I just noticed this. `ApplicationSystemTestCase` defines `driven_by` and is generated if you don't already have an app level system test case (`bin/rails g system_test xyz`). Can you explain why this was refactored to patch `ActionDispatch::SystemTestCase` directly?

Comment (ghiculescu) [reply guides/source/active_storage_overview.md:996]: Technically `ActionDispatch::SystemTestCase` defines `driven_by`, but point taken about the generator - I didn't know about that. I changed it to make the examples more consistent; I'll change the system ones back.

Comment (gmcgibbon) [reply guides/source/active_storage_overview.md]: Er, yes. My mistake, I meant `driven_by` is usually _called_ on the application test case.

Will merge now. Thanks for your patience on this!

**Comment by ghiculescu** [comment]
@gmcgibbon thanks so much for your help here! 🎉  🥇 💯
